### PR TITLE
21224 - Allow people to save avatar bookmarks

### PR DESCRIPTION
**Pull request by Triplelexx**
**Test Plan**
In addition to the "location" bookmarks, there should also now be an avatar bookmarking system presented in a similar fashion.

There should be the usual sub-menu under navigate for recalling bookmarks, there should also be a new 
sub-menu under avatar for recalling avatar bookmarks.

You should notice that the list is sorted.
Any existing bookmarks file you have saved should be parsed without modification.
Deleting the files should be tested, along with modifying the lists.
They should be located in the "local" folder.

The options are mostly the same for the menus:
You should be able to add an avatar bookmark, as well as delete one from the list.
When adding, a dialog should appear to enter the name.
If the name already exists a confirmation should be presented.
Delete should show the items in a combo box.

Only desktop/mouse operation has been tested.

**NB: In addition to the bookmark changes, some button functionality has been adapted.**
Buttons that must be tested are:
* The ATP asset "Choose File" button
* The running scripts "FROM DISK" button
* the running scripts "FROM URL" button

Additionally, there have been changes to the avatar settings dialog that should appear when you select settings -> avatar.

The "Appearance" input field should now validate the ".fst" extension.
The validation should let through extended extensions like putting a ? after.
The browse button should appear underneath on it's own without valid input.

If the ".fst" extension is detected, a bookmark button should then appear to the left of the browse button pushing it along.

The button should work the same as adding with the menu option.

**Known Issues**
* If the bookmarks lists are open when adding an entry the list will clear until returning.
* There have been recent changes to the tablet as a whole that were merged with this version.

 
.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 255e781cc7ca796e620ea84da54b86a99df17103 available at: https://gist.github.com/1589b2945db5b3eb3146e69d78538dcd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-255e781.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-255e781.dmg

**Comment by Nex-Pro** [comment]
Checking this.

**Comment by Nex-Pro** [comment]
#10046

Testing platform: Intel workstation.
OS: Win10 pro 64
CPU: i7 4.0 ghz
RAM: 16gb
GPU: Nvidia GTX1070 8gb VRam
HMD: Htc Vive.


IMO **PASS** but please read italicised note.

Test Plan
In addition to the "location" bookmarks,
there should also now be an avatar bookmarking system presented in a similar fashion.


There should be the usual sub-menu under navigate for recalling bookmarks, : **CHECK**
there should also be a new sub-menu under avatar for recalling avatar bookmarks: **SUCCESS**


You should notice that the list is sorted: **SUCCESS**
Any existing bookmarks file you have saved should be parsed without modification.
Deleting the files should be tested, along with modifying the lists: **SUCCESS**

***There is a JSON parsing issue whereby a dangling comma in the avatarbookmarks.json file will prevent the file loading. This is/should be allowed in js.***

They should be located in the "local" folder.


The options are mostly the same for the menus:
You should be able to add an avatar bookmark, as well as delete one from the list: **SUCCESS**
When adding, a dialog should appear to enter the name: **SUCCESS**
If the name already exists a confirmation should be presented: **SUCCESS**
Delete should show the items in a combo box: **SUCCESS**


Only desktop/mouse operation has been tested.


NB: In addition to the bookmark changes, some button functionality has been adapted.
Buttons that must be tested are:


The ATP asset "Choose File" button: **SUCCESS**
The running scripts "FROM DISK" button: **SUCCESS**
the running scripts "FROM URL" button: **SUCCESS**


Additionally, there have been changes to the avatar settings dialog
that should appear when you select settings -> avatar.

The "Appearance" input field should now validate the ".fst" extension: **SUCCESS**
The validation should let through extended extensions like putting a ? after: **SUCCESS**
The browse button should appear underneath on it's own without valid input: **SUCCESS**


If the ".fst" extension is detected,
a bookmark button should then appear to the left of the browse button pushing it along: **SUCCESS**


The button should work the same as adding with the menu option: **SUCCESS**


Known Issues

If the bookmarks lists are open when adding an entry the list will clear until returning.
There have been recent changes to the tablet as a whole that were merged with this version.
.


## QA Test
**QA Tester**: Piper.Peppercorn
**localhost 'createTestBlocks.js' Result**: SUCCESS
**localhost ATP Upload + Edit Result**: SUCCESS
**Echo Server Audio**: **Echo Server Audio** SUCCESS
**localhost 'harmonicOscillator.js' Result**: SUCCESS
**localhost 'reverbTest.js' Result**: SUCCESS
**localhost 'playaPerformanceTest.js' Result**: SUCCESS
**dev-playa 'playaPerformanceTest.js' Result**: SUCCESS

**QA Pass**: **QA Pass** SUCCESS
**Additional Test Results + Observations**:

Render rate and Present New Rate are shown incorrectly here
But were just on 90 there seems to be some issue with the test scripts.

**Domain** |localhost |dev-playa |
---|---|---|---
**Render Rate**: |0 |0 |
**Present Rate**: |90 |94 |
**Present New Rate**: |0 |90 |
**CPU Buffer Count**: |2041 |3940 |
**GPU Buffer Count**: |663 |1253 |
**CPU Memory Usage**: |23345296 |66364880 |
**GPU Memory Usage**: |14943664 |47433136 |
**CPU Texture Count**: |224 |233 |
**GPU Texture Count**: |158 |202 |
**GPU Transfer Count**: |0 |0 |
**CPU Texture Memory Usage**: |0 |0 |
**GPU Texture Memory Usage**: |596918553 |824856527 |
**GPU Virtual Memory Usage**: |0|0|
**Triangle Count**: |625636 |2479683 |
**Triangle Rate**: |62563600 |225425727 |
**API Drawcall Count**: |187 |351 |
**Drawcall Count**: |1174 |978 |
**Drawcall Rate**: |117400 |88909 |
**State Changes Texture Count**:|335 |318 |
**Pipeline Count**: |47 |42 |
**Item Opaques**: |229 |533 |
**Item Translucents**: |66 |26 |
**Item Lights**: |9 |49 |
**Timing Opaques**: |0.934 |2.561 |
**Timing Translucents**: |0.127 |0.076 |
**Timing Lighting**: |0.014 |0.021 |
**Render Frame**: |1.739 |4.005 |

**Comment by RyanDowne** [comment]
The issue I see with this is that when you are in the Settings/Avatar and click the save bookmark button, nothing happens. 

It really exciting though. This is really gonna save me time :)

**Comment by Triplelexx** [comment]
Thanks for the feedback.

@Nex-Pro Regarding the trailing comma / JSON parsing error I can have a look into that later today.
@RyanDowne That will be due to tablet changes, I'll try and get a handle on all the changes later, I was also waiting to see if anything else is observed. If you are in desktop mode with toolbar enabled it should work for you.

**Comment by cozza13** [comment]
ok to test

**Comment by hifi-gustavo** [comment]
Build console output for commit: 255e781cc7ca796e620ea84da54b86a99df17103 available at: https://gist.github.com/1589b2945db5b3eb3146e69d78538dcd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-255e781.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-255e781.dmg

**Comment by RyanDowne** [comment]
If we want to get this merged, I would suggest removing the button in the Settings/Avatar dialog. Or, if there is an easy way, make sure to invoke the other tablet screen when pressed.

**Comment by Triplelexx** [comment]
pushing to check for build errors, will look over again after break.

**Comment by hifi-gustavo** [comment]
Build console output for commit: ae3f04288d446ebc8b1091ff9e36b7ffe2ccf0a0 available at: https://gist.github.com/1589b2945db5b3eb3146e69d78538dcd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-ae3f042.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-ae3f042.dmg

File: interface/src/Bookmarks.cpp [hunk 1]
```diff
@@ -64,6 +114,40 @@ bool Bookmarks::contains(const QString& name) const {
...
+void Bookmarks::sortActions(Menu* menubar, MenuWrapper* menu) {
+    QList<QAction*> actions = menu->actions();
+    qSort(actions.begin(), actions.end(), sortOrder);
+    for (QAction* action : menu->actions()) {
+        menu->removeAction(action);
+        //removeBookmarkFromMenu(menubar, action->text());
+    }
+    for (QAction* action : actions) {
+        menu->addAction(action);
+//        for (int i = 0; i <  _bookmarks.size() ; i++) {
```
Comment (ZappoMan) [inline interface/src/Bookmarks.cpp]: Dead code?

File: interface/src/LocationBookmarks.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    QAction* teleportAction = _bookmarksMenu->newAction();
+    teleportAction->setData(address);
+    connect(teleportAction, SIGNAL(triggered()), this, SLOT(teleportToBookmark()));
+    if (!_isMenuSorted) {
+        menubar->addActionToQMenuAndActionHash(_bookmarksMenu, teleportAction, name, 0, QAction::NoRole);
+    } else {
+        // TODO: this is aggressive but other alternatives have proved less fruitful so far.
+
+        //menubar->addActionToQMenuAndActionHash(_bookmarksMenu, teleportAction, name, 0, QAction::NoRole, 
+        //Bookmarks::getMenuItemLocation(_bookmarksMenu->actions(), name));
```
Comment (ZappoMan) [inline interface/src/LocationBookmarks.cpp]: Dead code?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 57d6213175675b77e7cf8bf248b19e752372e297 available at: https://gist.github.com/1589b2945db5b3eb3146e69d78538dcd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-57d6213.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-57d6213.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8aa3e56ad459ccae10cc5f20230c68bbacdbf5c8 available at: https://gist.github.com/1589b2945db5b3eb3146e69d78538dcd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-8aa3e56.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-8aa3e56.dmg

**Comment by RyanDowne** [comment]
This is a QA Fail for me. 

Repro:
Open Marketplace
Select the stick and ball avatar
Open tablet in HMD/Settings/Avatar
Click add bookmark
name bookmark
Click OK
Crash

**Comment by Triplelexx** [comment]
@RyanDowne  @themelissabrown I don't know how I missed that crash when testing. I thought I fixed the problem, I was clearly mistaken. I'm running out of ideas to resolve the issue on my own.

**Comment by Nex-Pro** [comment]
@RyanDowne @Triplelexx Chief suspect in the crime = Stick and ball avatar.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 01b13d7032d51985c965bc14c362c43610d36724 available at: https://gist.github.com/1589b2945db5b3eb3146e69d78538dcd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-01b13d7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-01b13d7.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b4b5d4fe9bb8ee5675bf089be1ce03cb6d26bf77 available at: https://gist.github.com/1589b2945db5b3eb3146e69d78538dcd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-b4b5d4f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-b4b5d4f.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: af95741a7e52405539cc4a8445e06c227c52cb23 available at: https://gist.github.com/1589b2945db5b3eb3146e69d78538dcd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-af95741.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-af95741.dmg

**Comment by themelissabrown** [comment]
Please test again with stick and ball avatar and without to see if the crash is specific to that avatar.

**Comment by RyanDowne** [comment]
Re-tested and the app no longer crashes when you save a bookmark from the settings menu

**Comment by claudioshg** [comment]
testing

**Comment by Nex-Pro** [comment]
@claudioshg Chris has requested me to test ths

**Comment by Nex-Pro** [comment]
I don't know why my previous comment didn't show up

**Comment by claudioshg** [comment]
Testing

**Comment by claudioshg** [comment]
There was a power outage caused by high winds where I live. That's the reason for the 8-hour delay on this PR. I will resume testing immediately.

**Comment by claudioshg** [comment]
### Desktop PC - Windows 10 Home - CPU 4.0GHz Intel i7 - GTX1070 - RAM 16GB - Tested with Oculus Rift

Tested: Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-af95741.exe


All testing completed in HMD

- There should be the usual sub-menu under navigate for recalling bookmarks, there should also be a new
sub-menu under avatar for recalling avatar bookmarks.
SUCCESS
- You should notice that the list is sorted.
SUCCESS

NOTE:
Avatar Settings> Browse Avatars> Click on any Avatar> Once in Avatars page, no way to return to previous page.
- Any existing bookmarks file you have saved should be parsed without modification.
PASS: Seems like they are listed in alphabetical order.
- Deleting the files should be tested
PASS
- along with modifying the lists.
PASS
- They should be located in the "local" folder.
PASS: I found avatarbookmarks.json in \AppData\Local\High Fidelity - PR10046\Interface
- The options are mostly the same for the menus:
You should be able to add an avatar bookmark, as well as delete one from the list.
PASS
- When adding, a dialog should appear to enter the name.
PASS
- If the name already exists a confirmation should be presented.
PASS
- Delete should show the items in a combo box.
PASS
- Buttons that must be tested are: The ATP asset "Choose File" button
PASS: Clicking on "Choose File" opens 'Modal' dialog.

NOTE: Asset Directory> Choose File> Clicking on 'X'/HOME button does nothing. 
I had two crashes while clicking on the 'X'/HOME button on the tablet. Attempted clicking with stylus/finger/mouse/laser.
- Buttons that must be tested are: The running scripts "FROM DISK" button
PASS: Clicking on "From Disk" opens 'Open Script' dialog.

NOTE: MENU> Edit> Running Scripts...> FROM DISK> Clicking on 'X'/HOME button does nothing.
Attempted clicking on 'X' button with stylus/finger/mouse/laser.
- Buttons that must be tested are: The running scripts "FROM URL" button
PASS: Clicking on "From URL" opens 'Open and Run Script' dialog.

NOTE: MENU> Edit> Running Scripts...> FROM URL> Clicking on 'X'/HOME button does nothing.
Attempted clicking on 'X' button with stylus/finger/mouse/laser.
- The "Appearance" input field should now validate the ".fst" extension.
PASS
- The validation should let through extended extensions like putting a ? after.
PASS
- The browse button should appear underneath on it's own without valid input.
PASS
- If the ".fst" extension is detected, a bookmark button should then appear to the left of the browse button pushing it along.
PASS
- The button should work the same as adding with the menu option.
PASS

File: interface/src/Bookmarks.cpp [hunk 3]
```diff
@@ -27,15 +23,69 @@
...
+void Bookmarks::addBookmarkToFile(const QString& bookmarkName, const QString& bookmarkAddress) {
+    Menu* menubar = Menu::getInstance();
+    if (contains(bookmarkName)) {
+        auto offscreenUi = DependencyManager::get<OffscreenUi>();
+        auto duplicateBookmarkMessage = offscreenUi->createMessageBox(OffscreenUi::ICON_WARNING, "Duplicate Bookmark",
+            "The bookmark name you entered already exists in your list.",
+            QMessageBox::Yes | QMessageBox::No, QMessageBox::Yes);
+        duplicateBookmarkMessage->setProperty("informativeText", "Would you like to overwrite it?");
+
+        auto result = offscreenUi->waitForMessageBoxResult(duplicateBookmarkMessage);
```
Comment (ZappoMan) [inline interface/src/Bookmarks.cpp]: I am worried that this blocking message box will not work in HMD mode with lasers... have we tested that?

File: interface/src/Bookmarks.cpp [hunk 4]
```diff
@@ -27,15 +23,69 @@
...
+        QString bookmarkName = it.key();
+        QString bookmarkAddress = it.value().toString();
+        addBookmarkToMenu(menubar, bookmarkName, bookmarkAddress);
+    }
+}
+
+void Bookmarks::deleteBookmark() {
+    QStringList bookmarkList;
+    QList<QAction*> menuItems = _bookmarksMenu->actions();
+    for (int i = 0; i < menuItems.count(); i += 1) {
```
Comment (ZappoMan) [inline interface/src/Bookmarks.cpp]: should be `i++`

**Comment by ZappoMan** [comment]
@Triplelexx @Nex-Pro @RyanDowne - can we please retest this with the following path:
- add a bookmark with name "foo" (or your choice)
- then add a second bookmark with the same name "foo" (or your choice)
- from the code it appears that this pops up a message box asking to confirm...
- *Please test this with HMD + hand controllers + lasers*... make sure the lasers still work when the message box is up.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4c2823b42df7a724e8bd978be27c10f7680cba6c available at: https://gist.github.com/1589b2945db5b3eb3146e69d78538dcd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-4c2823b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10046-4c2823b.dmg

**Comment by Nex-Pro** [comment]
I'll get on this ASAIC

**Comment by Nex-Pro** [comment]
Testing this

**Comment by Nex-Pro** [comment]
### #10046 

Testing platform: Intel workstation.
OS: Win10 pro 64
CPU: i7 4.0 ghz
RAM: 16gb
GPU: Nvidia GTX1070 8gb VRam
HMD: Htc Vive.

## PASS

In addition to the "location" bookmarks, there should also now be an avatar bookmarking system presented in a similar fashion. **SUCCESS**

There should be the usual sub-menu under navigate for recalling bookmarks, there should also be a new sub-menu under avatar for recalling avatar bookmarks.  **SUCCESS**

You should notice that the list is sorted. **SUCCESS**

Any existing bookmarks file you have saved should be parsed without modification.
Deleting the files should be tested, along with modifying the lists. **SUCCESS for delete and mod**
They should be located in the "local" folder.


The options are mostly the same for the menus:
You should be able to add an avatar bookmark, as well as delete one from the list. **SUCCESS**
When adding, a dialog should appear to enter the name. **SUCCESS**
If the name already exists a confirmation should be presented. **SUCCESS**
Delete should show the items in a combo box. **SUCCESS**
***Needs a redesign.  Suggest checkboxes and a delete button with subsequent confirm/cancel on list of bookmarks.***

Only desktop/mouse operation has been tested. **Tested in Vive**

NB: In addition to the bookmark changes, some button functionality has been adapted.
Buttons that must be tested are:

The ATP asset "Choose File" button **SUCCESS**

The running scripts "FROM DISK" button **SUCCESS**

the running scripts "FROM URL" button **SUCCESS**


Additionally, there have been changes to the avatar settings dialog that should appear when you select settings -> avatar.

The "Appearance" input field should now validate the ".fst" extension. **SUCCESS**

The validation should let through extended extensions like putting a ? after. **SUCCESS Parsed with cachebuster**

The browse button should appear underneath on it's own without valid input. **SUCCESS**


If the ".fst" extension is detected, a bookmark button should then appear to the left of the browse button pushing it along. **SUCCESS**

The button should work the same as adding with the menu option. **SUCCESS** 

Known Issues

If the bookmarks lists are open when adding an entry the list will clear until returning.
There have been recent changes to the tablet as a whole that were merged with this version.


Notes: Got a crash ( splat sent ) while using the [avatar bookmark] button in the avatar appearance dialog.

Also ( good point ) I tested removing the url to my currently worn avatar in the appearance dialog. 
Then before saving I used the [avatar bookmark] button. The current avatar was successfully bookmarked
and on returning to the appearance dialog the URL was restored.
See my note (further up page) re. the 'delete bookmark' design.

@Triplelexx  Nice clear notes, thanks.

**Comment by RyanDowne** [comment]
Is this dev approved too?

**Comment by themelissabrown** [comment]
I haven't gotten dev approval yet. Will pester again

File: interface/src/AvatarBookmarks.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+#include <avatar/AvatarManager.h>
+
+#include "MainWindow.h"
+#include "Menu.h"
+
+#include "AvatarBookmarks.h"
+#include <QtQuick/QQuickWindow>
+
+AvatarBookmarks::AvatarBookmarks() {
+    _bookmarksFilename = QStandardPaths::writableLocation(QStandardPaths::DataLocation) + "/" + AVATARBOOKMARKS_FILENAME;
```
Comment (zfox23) [inline interface/src/AvatarBookmarks.cpp]: This seems like a strange path for this file. Why not use the same path as the `Interface.json` settings file?

Comment (Triplelexx) [inline interface/src/AvatarBookmarks.cpp]: This is the location that bookmark were stored in prior to my changes. I agree it should be AppDataLocation ideally.

Comment (zfox23) [inline interface/src/AvatarBookmarks.cpp]: OK

**Pull request merged**